shell/common/rasterizer.cc

```diff
@@ -536,60 +536,60 @@ RasterStatus Rasterizer::DrawToSurfaceUnsafe(
-
-    // Disable partial repaint if external_view_embedder_ SubmitFrame is
-    // involved - ExternalViewEmbedder unconditionally clears the entire
-    // surface and also partial repaint with platform view present is something
-    // that still need to be figured out.
-    bool disable_partial_repaint =
-        external_view_embedder_ &&
-        (!raster_thread_merger_ || raster_thread_merger_->IsMerged());
-
-    FrameDamage damage;
-    if (!disable_partial_repaint && frame->framebuffer_info().existing_damage) {
-      damage.SetPreviousLayerTree(last_layer_tree_.get());
-      damage.AddAdditonalDamage(*frame->framebuffer_info().existing_damage);
-    }
+  if (!compositor_frame) {
+    return RasterStatus::kFailed;
+  }
 
-    RasterStatus raster_status =
-        compositor_frame->Raster(layer_tree, false, &damage);
-    if (raster_status == RasterStatus::kFailed ||
-        raster_status == RasterStatus::kSkipAndRetry) {
-      return raster_status;
-    }
+  compositor_context_->raster_cache().PrepareNewFrame();
+  frame_timings_recorder.RecordRasterStart(fml::TimePoint::Now());
 
-    SurfaceFrame::SubmitInfo submit_info;
-    submit_info.frame_damage = damage.GetFrameDamage();
-    submit_info.buffer_damage = damage.GetBufferDamage();
+  // Disable partial repaint if external_view_embedder_ SubmitFrame is
+  // involved - ExternalViewEmbedder unconditionally clears the entire
+  // surface and also partial repaint with platform view present is something
+  // that still need to be figured out.
+  bool disable_partial_repaint =
+      external_view_embedder_ &&
+      (!raster_thread_merger_ || raster_thread_merger_->IsMerged());
 
-    frame->set_submit_info(submit_info);
+  FrameDamage damage;
+  if (!disable_partial_repaint && frame->framebuffer_info().existing_damage) {
+    damage.SetPreviousLayerTree(last_layer_tree_.get());
+    damage.AddAdditonalDamage(*frame->framebuffer_info().existing_damage);
+  }
 
-    if (external_view_embedder_ &&
-        (!raster_thread_merger_ || raster_thread_merger_->IsMerged())) {
-      FML_DCHECK(!frame->IsSubmitted());
-      external_view_embedder_->SubmitFrame(surface_->GetContext(),
-                                           std::move(frame));
-    } else {
-      frame->Submit();
-    }
+  RasterStatus raster_status =
+      compositor_frame->Raster(layer_tree, false, &damage);
+  if (raster_status == RasterStatus::kFailed ||
+      raster_status == RasterStatus::kSkipAndRetry) {
+    return raster_status;
+  }
 
-    compositor_context_->raster_cache().CleanupAfterFrame();
-    frame_timings_recorder.RecordRasterEnd(
-        &compositor_context_->raster_cache());
-    FireNextFrameCallbackIfPresent();
+  SurfaceFrame::SubmitInfo submit_info;
+  submit_info.frame_damage = damage.GetFrameDamage();
+  submit_info.buffer_damage = damage.GetBufferDamage();
+  submit_info.target_time = frame_timings_recorder.GetVsyncTargetTime();
```

Comment:
Since we're now depending on GetVsyncTargetTime for something other than metrics, I think we'll need to update VsyncWaiter to always get the current Display refresh rate instead of caching it. https://github.com/flutter/engine/blob/main/shell/platform/android/io/flutter/view/VsyncWaiter.java#L39

On Android S and later, `Display.getRefreshRate()` returns the current display refresh rate, instead of the app maximum. https://developer.android.com/about/versions/12/reference/compat-framework-changes#display_info_nr_advanced_supported

Reply:
@jason-simmons FYI - since I think you recently changed this to _not_ do that.

Reply:
The call to `getRefreshRate()` was taking a significant amount of time on the platform thread in some cases (see https://github.com/flutter/flutter/issues/90883)

Reintroducing that call may be risky if it needs to be done on every frame.

Reply:
Do we know if those devices were susceptible when running on Android S?

I guess the problem here is that some devices will scale the refresh rate automatically/dynamically.

It might also be interesting to see if the same delays exist when using the NDK API instead.

Reply:
The Huawei device referenced in https://github.com/flutter/flutter/issues/90883#issuecomment-930682438 was running Android P

I don't know if Android S ensures better performance for getRefreshRate

Reply:
https://github.com/flutter/engine/pull/29800 Should resolve this.